c7n/resources/s3.py

```diff
@@ -1726,3 +1727,110 @@ def process_chunk(self, batch, bucket):
             Bucket=bucket['Name'], Delete={'Objects': objects}).get('Deleted', ())
         if self.get_bucket_style(bucket) != 'versioned':
             return results
+
+
+@actions.register('lifecycle')
+class Lifecycle(BucketActionBase):
+    """Action applies a lifecycle policy to versioned S3 buckets
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: s3-apply-lifecycle
+                resource: s3
+                actions:
+                  - type: lifecycle
+                    id: default-retention-policy
+                    expire-days: 60
+                    expire-obj-del-marker: false
+                    noncur-expire-days: 60
+    """
+
+    schema = type_schema(
+        'lifecycle',
+        **{
+            'id': {'type': 'string'},
```

Comment:
if this is required should pass in required keyword set to list of keys that must be defined.